Wikipedia/Code/WMFTaskGroup.m

```diff
@@ -0,0 +1,50 @@
+#import "WMFTaskGroup.h"
+
+@interface WMFTaskGroup ()
+    
+@property (nonatomic) dispatch_group_t group;
+@property (nonatomic) NSInteger count;
+
+@end
+
+@implementation WMFTaskGroup
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        self.group = dispatch_group_create();
+    }
+    return self;
+}
+- (void)enter {
+    @synchronized (self) {
+        self.count++;
+        dispatch_group_enter(_group);
+    }
+}
+
+- (void)leave {
+    @synchronized (self) {
+        if (self.count > 0) {
+            self.count--;
+            dispatch_group_leave(self.group);
+        } else {
+            DDLogError(@"Mismatched leave for group: %@", self);
+        }
+    }
+}
+
+- (void)waitInBackgroundWithTimeout:(NSTimeInterval)timeout completion:(nonnull dispatch_block_t)completion {
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
```

Comment:
@coreyfloyd i think that's being used on line 47. this one is so you can also have a timeout on the group independent of the tasks in the group, right?